examples/userapi/04_boundary_conditions.ipynb

```diff
@@ -0,0 +1,471 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Boundary conditions tutorial"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This tutorial aims to demonstrate how users can implement various boundary conditions in Devito, building on concepts introduced in previous tutorials. Over the course of this notebook we will go over the implementation of both free surface boundary conditions and perfectly-matched layers (PMLs) in the context of the first-order acoustic wave equation. This tutorial is based on a simplified version of the method outlined in Liu and Tao's 1997 paper (https://doi.org/10.1121/1.419657).\n",
+    "\n",
+    "For purely educational purposes, we will set up our domain with PMLs along the left and right edges, and free surface boundaries at the top and bottom.\n",
+    "\n",
+    "(Add a picture here)\n",
+    "\n",
+    "As always, we will begin by specifying some parameters for our `Grid`:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "shape = (101, 101)\n",
+    "extent = (2000., 2000.)\n",
+    "PMLS = 10  # Number of PMLs on left and right sides"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Additionally, we will need to specify the dimensions for our `Grid`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from devito import SpaceDimension, Constant\n",
+    "\n",
+    "x = SpaceDimension(name='x', spacing=Constant(name='h_x',\n",
+    "                                              value=extent[0]/(shape[0]-1)))\n",
+    "y = SpaceDimension(name='y', spacing=Constant(name='h_y',\n",
+    "                                              value=extent[1]/(shape[1]-1)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We will need to use subdomains to accomodate the modified equations in the PML regions. As `Grid` objects are immutable, we must define our subdomains beforehand."
```

Comment:
I wouldn't talk about "immutability", this will only confuse people. Just describe the API. In fact, you can probably just drop the sentence "As Grid objects are immutable, ... beforehand."

Reply:
I made this more euphemistic, but left it in in spirit. "Immutibility" isn't mentioned anymore, but I've kept a bit about how subdomains come before grid, as the order is a tad unintuitive for new users otherwise